README.md
#Full Page Navigation by Pete R.
Create a fully responsive and dynamic full page navigation with one JS call
Created by [Pete R.](http://www.thepetedesign.com), Founder of [BucketListly](http://www.bucketlistly.com)

[![Full Page Navigation](http://www.thepetedesign.com/images/fullpagenav_image.png "Full Page Navigation")](http://www.thepetedesign.com/demos/fullpagenav_demo.html)

## Demo
[View demo](http://www.thepetedesign.com/demos/fullpagenav_demo.html)

## Compatibility
Modern browsers such as Chrome, Firefox, and Safari on both desktop and smartphones have been tested. I have not test this on IE.

## Basic Usage
Full Page Navigation is a simple script that let you create a full page navigation (navigation shown as columns) dynamically with one JS call. You can define how your navigation will be divided, how fast/slow the animation will be and many more features.

To add this to your website, simply include the latest jQuery library together with `jquery.fullpagenav.js`, and `fullpagenav.css` into your document's `<head>` and create an HTML markup as follows:

````html
<body>
...
  <ul class="nav">
    <li> ... </li>
    <li> ... </li>
    <li> ... </li>
    ...
  </ul>
...
</body>
````

And you can add any kind of HTML tags that you wish for it to display in each column inside the `<li>` tag. You can use different tags if you wish. You will need to assign a new selector though. More info on that below. Once this is done, call the script like this:
 
````javascript
$(".nav").fullpagenav({
  columns: 5, // Number of columns you wish the script to divide your navigation into. This should be relative to the number of li tags you have. The column should be 3 for the example above. The default value is 5.
  selector: "> li", // In case you don't want to use ul and li, you can change the selector here. This option accepts CSS selector. The Default value is "> li"
  hoverSize: "30%", // This option let you determine how big to expand the navigation when the cursor is hovering the area. The option accepts a percentage value. The default value is "30%"
  animateDuration: 500, // You can define the animation duration here which will be used globally throughout the plugin. This option accepts milliseconds without the unit. The default value is 500.
  animateFrom: "left", // Determine which direction the hovered navigation should expand to. Available options are "left", "right", and "auto". Auto will determine the direction from where you cursor enters the navigation area. The default value is "left".
  clickable: true, // Toggle this to true if you want the area to be clickable. Make sure to add an extra data-link attribute to the li tag containing the URL you wish the navigation to take to user to once clicked. The default value is true.
  afterClicked: null // A callback which will be fired after the user clicked the navigation
});
````

## Making it Clickable
Here's an example on how to make the navigation clickable. Make sure you have the HTML markup as follows:

````html
<body>
...
  <ul class="nav">
    <li data-link="http://www.bucketlistly.com"> ... </li>
    <li data-link="http://www.bucketlistly.com"> ... </li>
    <li data-link="http://www.bucketlistly.com"> ... </li>
    ...
  </ul>
...
</body>
```` 
Note: Don't forget to change the data-link attribute to the URL you prefer. 

Now, simply toggle the clickable option to true like this.

````javascript
$(".nav").fullpagenav({
  clickable: true
});
```` 
And the script will automatically make the whole area clickable, with a nice animation expanding the clicked area to full page and load the URL.

## Callbacks
You can use this callback to perform actions after you clicked the navigation. This is perfect for when you want to populate the page with an ajax instead of a traditional page load.

### afterClicked(URL)
This callback gets called after the user clicked the navigation.

````javascript
  $(".main").fullpagenav({
    afterClicked: function(link) {
      ...
    }
  });
````

And that's a wrap for all the features available for this plugin. If you want to see more of my plugins, visit [The Pete Design](http://www.thepetedesign.com/#design), or follow me on [Twitter](http://www.twitter.com/peachananr) and [Github](http://www.github.com/peachananr).

## Other Resources
- [Tutorial](http://www.onextrapixel.com/2013/12/04/turn-your-website-into-a-fully-dynamic-full-page-column-based-navigation/)


CHANGELOG
v2 in Development
  - Change click behavior to expand the slide.
  - Fixed mouse hover to transition from one state to the next fluidly.
  - Added fullpagenav(n) to set the expanded page.
  - Added fullpagenav("prev") to go backwards through the pages.
  - Added fullpagenav("next") to forward through the pages.
  - Removed the data-link attribute.
  - Added events expanding.fullpagenav, shrinking.fullpagenav,
    expanded.fullpagenav
  - Added classes for state expanded, expanding.
  - Cleaned up all the code and organized it similarly to bootstrap plugins.
  - Only expose one jQuery method.
  - Added page transitions between full pages.
  - Added html attribute data-fullpagenav-to for navigating between views.
  - Added html attribute data-fullpagenav-direction for specifying nav direction.
  - Allow highlighting to be caught through overlays.
  - Added data attribute data-fullpagenav-width for individual column widths.


jquery.fullpagenav.js
/* ===========================================================
 * jquery-fullpagenav.js v1
 * ===========================================================
 * Copyright 2013 Pete Rojwongsuriya.
 * http://www.thepetedesign.com
 *
 * Divide animated navigation into columns
 * filling up the whole page with a simple JS call
 *
 * https://github.com/peachananr/fullpagenav
 *
 * ========================================================== */

(function($) {
  "use strict";

  var FullPageNav = function(element, options) {
    var that = this;
    this.$element = $(element);
    this.options = options;

    this.$items = this.$element.find(this.options.selector);
    this.$items.wrapInner("<div class='fpn_wrap'>");

    // @todo There's a CSS attribute that allows mouse events to flow through
    // which isn't widely supported on IE. On mot other browsers this
    // shouldn't be necessary.
    // We should probably introduce some option like
    // catchMoseEvents: 'css-unsupported' (default) / 'always' / false
    $("body").on("mousemove", function(e) {
      var inFocus = that.$items.filter(function(index, item) {
        var bounds = item.getBoundingClientRect();

        if(bounds.left > e.clientX || bounds.right < e.clientX ||
          bounds.top > e.clientY || bounds.bottom < e.clientY){
          return;
        }
        return true;
      });

      if(!inFocus.is(".highlight")){
        that.$items.filter(".highlight").trigger("mouseleave");
        inFocus.trigger("mouseenter");
      }
    });

    this.$element.on("mouseenter",
      this.options.selector,
      this.highlight.bind(this));

    this.$element.on("mouseleave",
      this.options.selector,
      this.unhighlight.bind(this));


    this.$element.on("click", this.options.selector, function(e) {
      e.preventDefault();
      if(!that.$items.hasClass("active")){
        that.show($(this));
      }
    });

    this.reflow(true);
  };

  FullPageNav.DEFAULTS = {
    selector: "> li",
    hoverSize: 30,
    animateDuration: 500,
    easing: "linear",
    animateFrom: "left"
  };

  FullPageNav.prototype.highlight = function(e) {
    if(this.$items.hasClass("active")){
      return;
    }
    this.$items.filter(".highlight").removeClass("highlight");
    $(e.currentTarget).addClass("highlight");

    this.reflow();
  };

  FullPageNav.prototype.unhighlight = function(e) {
    if(this.$items.hasClass("active") || !this.$items.hasClass("highlight")){
      return;
    }
    $(e.currentTarget).removeClass("highlight");
    this.reflow();
  };

  FullPageNav.prototype.getItemIndex = function(item) {
    this.$items = item.parent().children(this.options.selector);
    return this.$items.index(item || this.$active);
  };

  FullPageNav.prototype.next = function() {
    return this.show("next");
  };

  FullPageNav.prototype.prev = function() {
    return this.show("prev");
  };

  FullPageNav.prototype.show = function(next) {
    var direction = next.direction;
    var $next = typeof next.show === 'function' ? next : next.show;
    var $active = $(this.options.selector, this.$element).filter(".active");

    if($active.length && !direction){
      direction = this.$items.index($active) < this.$items.index($next) ?
        "left" : "right";
    }
    var that = this;

    if(!$next.length){
      return this;
    }

    if($next.hasClass("active")){
      return this;
    }

    this.$element.addClass("fullpage");

    var relatedTarget = $next[0];

    var showEvent = $.Event("show.fullpagenav", {
      relatedTarget: relatedTarget
    });
    this.$element.trigger(showEvent);

    var shownEvent = $.Event("shown.fullpagenav", {
      relatedTarget: relatedTarget
    });

    $next.addClass("active");
    $active.removeClass("active");

    if($active.length){
      $next.css("left", direction === 'left' ? "100%" : "-100%");
      $active
        .add($next)
        .addClass(direction + " animate")
        .animate({left: (direction === 'left'? '-' : '+') + '=100%'},
        that.options.animateDuration,
        that.options.easing);
      $next.promise().done(function() {
        $active.add($next).removeClass("animate left right");
        that.$element.trigger(shownEvent);
      });
    }else{
      this.reflow().done(function() {
        that.$element.trigger(shownEvent);
        that.$items.attr("style", null);
      });
    }
    return this;
  };

  FullPageNav.prototype.close = function() {
    var $active = $(this.options.selector, this.$element).filter(".active");
    var that = this;
    if(!$active.length){
      return this;
    }
    $active
      .removeClass("active")
      .css({width: "100%", left: "0%"})
      .prevAll("li").css("left", "0%").end()
      .nextAll("li").css("left", "100%");

    this.$element.removeClass("fullpage");

    var collapseEvent = $.Event("collapse.fullpagenav");
    this.$element.trigger(collapseEvent);

    var collapsedEvent = $.Event("collapsed.fullpagenav");

    this.reflow().done(function() {
      $(that).trigger(collapsedEvent);
    });
    return this;
  };

  FullPageNav.prototype.to = function(pos) {
    this.$active = this.$element.find(".active");

    if(pos > (this.$items.length - 1) || pos < 0){
      return this;
    }

    return this.show(this.$items.eq(pos));
  };

  FullPageNav.prototype.reflow = function(instantly) {
    var options = this.options;
    var primaryItemWidth;
    var itemWidth;
    var offset = 0;
    var that = this;
    var expanding = this.$items.hasClass("active");

    if(!expanding){
      var remainingWidth = 100;
      var normal = 0;

      this.$items.each(function() {
        if($(this).hasClass("highlight")){
          remainingWidth -= that.options.hoverSize;
        }else if($(this).data('fullpagenav-width')){
          remainingWidth -= $(this).data('fullpagenav-width');
        }else{
          normal++;
        }
      });
      itemWidth = remainingWidth / normal;
    }

    var animations = this.$items.map(function() {
      var targetWidth = 0;
      var $item = $(this);
      if(expanding){
        targetWidth = $item.hasClass("active") ? 100 : 0;
      }else{
        if($item.hasClass("highlight")){
          targetWidth = that.options.hoverSize;
        }else{
          targetWidth = $item.data("fullpagenav-width") || itemWidth;
        }
      }

      var animation = $item.stop().animate({
        left: offset + "%",
        width: targetWidth + "%"
      }, options.animateDuration, options.easing);
      offset += targetWidth;
      return animation;
    });

    var promise = $.when.apply(this, animations);
    if(instantly){
      $.each(animations, function(){this.finish()});
    }
    return promise;
  };

  function Plugin(option, actionOption) {
    return this.each(function() {
      var $this = $(this);
      var data = $this.data("fullpagenav");
      var options = $.extend({}, FullPageNav.DEFAULTS, $this.data(),
          typeof option === "object" && option);

      var action = typeof option === "string" ? option : "reflow";

      if(!data) {
        $this.data("fullpagenav", (data = new FullPageNav(this, options)));
      }
      if(typeof option === "number"){
        data.to(option);
      }else{
        data[action](actionOption);
      }
    });
  }

  $(document).on('click', '[data-fullpagenav-to]', function(e) {
    var $this = $(this);
    var $target = $($this.data('fullpagenav-target'));
    if(!$target.hasClass('fullpagenav')) return;

    Plugin.apply($target, ["show", {
      show: $($this.data('fullpagenav-to')),
      direction: $this.data('fullpagenav-direction')
    }
    ]);

    e.preventDefault();
  });

  $.fn.fullpagenav = Plugin;
})(window.jQuery);
